Move native credentials lookup to the client library.

calavera opened the pull request:
This change extracts the native client logic from Docker
to use the one provided by the credentials library.

Tests remain here to verify that Docker's behavior with email
and whatnot still works, but the library also implements its own tests.

It would be awesome if someone could also review the changes in the library.
They should not affect the current behavior, but tests remain here to verify it.

https://github.com/docker/docker-credential-helpers/pull/15

Signed-off-by: David Calavera david.calavera@gmail.com

calavera commented:
Oh, if you wonder why I want to do this, it's because I'm going to use that code in another CLI :grin:

thaJeztah commented on hack/vendor.sh:
```diff
@@ -85,6 +85,7 @@ clone git github.com/ugorji/go 5abd4e96a45c386928ed2ca2a7ef63e2533e18ec
 clone git github.com/hashicorp/consul v0.5.2
 clone git github.com/boltdb/bolt v1.2.1
 clone git github.com/miekg/dns 75e6e86cc601825c5dbcd4e0c209eab180997cd7
+clone git github.com/docker/docker-credential-helpers 2ccf72627d3d89732442be0de61514c5f30701df https://github.com/calavera/docker-credential-helpers.git
```
Should we create a tag for this?

calavera replied:
first, I needs to land in master, it's in my branch waiting for a review :smile:

thaJeztah replied:
oh! you're right, ignore ignore :)

calavera commented:
Janky looks happy with the changes, but vendor will complain until docker/docker-credential-helpers#15 is merged. I do not have creds in that repo anymore :pensive:

thaJeztah commented:
@calavera before we merge this, we should have a "maintainers" file added to that repository, in case changes are needed that we have enough people to make those go in smoothly

runcom commented on vendor/src/github.com/docker/docker-credential-helpers/client/client.go:
```diff
@@ -0,0 +1,72 @@
+		t := strings.TrimSpace(string(out))
+
+		// do not return an error if the credentials are not
+		// in the keyckain. Let the caller ask for new credentials.
```
this comment is wrong now

vdemeester commented:
@calavera you can update the vendor now :wink:

calavera committed:
Vendor docker-credential-helpers v0.3.0

Fix a bug in the vendor helpers that took packages
that started with github.com/docker/docker like if
they were from within the project.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera committed:
Use docker-credential-helpers client to talk with native creds stores.

Signed-off-by: David Calavera <david.calavera@gmail.com>

calavera commented:
I finally figured out what was going on with the vendor script, it's a bug in `hack/.vendor-helpers`. It assumes that every package that starts with `github.com/docker/docker` belongs to the project and it was removing `github.com/docker/docker-credential-helpers` from the imported list. 

I included a patch in the vendoring commit:

https://github.com/docker/docker/pull/22988/files#diff-04aeacba9949f4aa0a8c6c812287ae5a

thaJeztah commented:
Ah! Thanks for fixing that!

thaJeztah commented:
LGTM

runcom commented on cliconfig/credentials/native_store.go:
```diff
@@ -127,70 +97,30 @@ func (c *nativeStore) storeCredentialsInStore(config types.AuthConfig) error {
-		if t == errCredentialsNotFound.Error() {
+		if credentials.IsErrCredentialsNotFound(err) {
+			// do not return an error if the credentials are not
+			// in the keyckain. Let docker ask for new credentials.
```
I think the caller should decide here if not found should be ignored and err not found should be returned along with ret. But this is not blocking

runcom commented:
LGTM

vdemeester commented:
LGTM 🐹
